### PR TITLE
Begin the effort of improving Chapel's first-class-function support

### PR DESCRIPTION
**Pull request by dlongnecke-cray**
Begin the effort of improving Chapel's first-class-function support

Add support for constructing function types and new syntax for
creating anonymous functions.

```chapel
type F = proc(int, int): int;
const adder: F = proc(x: int, y: int): int { return x + y; };
var num = adder(4, 4);
assert(num == 8);
```

This PR makes adjustments to the Chapel grammar. It introduces new
AST to `dyno` and improves the production compiler's representation
of function types. It also refactors the code which creates FCF
wrappers.

[Changes follow.]

Adjust the old compiler to not insert a RVV or return statement into
extern functions or functions marked with `FLAG_NO_FN_BODY`. This has
consequences for later passes which rely on those injected symbols
(e.g., garbage collection performed in prune), so make adjustments
as necessary.

Migrate the code which generates FCF wrappers and reasons about FCF
types to a new source file (with a corresponding header) called
`firstClassFunctions.cpp`.

Add new UAST nodes called `FunctionSignature` and `AnonFormal`, and
use them in the parser to store the result of parsing a function
type.

TESTING

- [x] `ALL` on `linux64`, `standard`, `-futures`
- [x] `ALL` on `linux64`, `COMM=gasnet`

Reviewed by @DanilaFe, @benharsh. Thanks!

FUTURE WORK

- This PR opts to lean on the existing FCF wrapper class types
  instead of introducing new AST/Type to the production compiler.
  These new types should be added in the future in order to
  model new semantics that can't be handled by the current
  class type representation (e.g., exported functions accepting a
  function value). This does not mean that the current
  effort is wasted - it can continue to be used for closure types.
  I expect that the new IR will look like 'TypedFnSignature' from
  the dyno compiler
- As part of this effort, will have to adjust all the necessary
  old AST to hash by value equivalence and not identity
- Clean up and generalize the algorithm used to compute the closure
  for nested functions. Right now it mutates the AST directly, and it
  would be nice if it would just compute some summary info

Signed-off-by: David Longnecker <dlongnecke-cray@users.noreply.github.com>

### CODE REVIEW

**Comment by dlongnecke-cray** [comment]
I expect the CI checks to fail until I rebuild the parser.

File: compiler/dyno/include/chpl/uast/Function.h
Comment (DanilaFe) [inline compiler/dyno/include/chpl/uast/Function.h:224]: Just to confirm: no function in Chapel code can ever be called `proc` since `proc` is a keyword, so this comparison is fine?

File: compiler/dyno/include/chpl/uast/Formal.h
Comment (DanilaFe) [inline compiler/dyno/include/chpl/uast/Formal.h:101]: Where is this implemented? I don't see it anywhere, but I do see a call to it in `AnonFormal` above.

**Review by DanilaFe — COMMENTED** [review]
Here's as far as I got on my first go. Probably lots of silly / obvious questions, but maybe those will help battle-test the PR :-)

File: compiler/AST/build.cpp
Comment (dlongnecke-cray) [reply compiler/AST/build.cpp:1899]: I removed it because we added it to dyno.

File: compiler/dyno/include/chpl/uast/AnonFormal.h
Comment (dlongnecke-cray) [reply compiler/dyno/include/chpl/uast/AnonFormal.h:51]: I'll drop the "call" and just say "resolution".

File: compiler/dyno/include/chpl/uast/AnonFormal.h
Comment (dlongnecke-cray) [reply compiler/dyno/include/chpl/uast/AnonFormal.h:74]: Thanks! Should be `int8_t`.

File: compiler/codegen/cg-expr.cpp
Comment (dlongnecke-cray) [reply compiler/codegen/cg-expr.cpp:3746]: Ok, ok, I'll revert it xD

File: compiler/dyno/include/chpl/uast/Function.h
Comment (dlongnecke-cray) [reply compiler/dyno/include/chpl/uast/Function.h:224]: Yeah, that's the idea behind this shortcut. We could also add different "Function::Kind" as well if we find this distasteful, I suppose. Or a `bool is Anonymous` field.

File: compiler/dyno/include/chpl/uast/Formal.h
Comment (dlongnecke-cray) [reply compiler/dyno/include/chpl/uast/Formal.h:101]: Wow, weird...why don't I get a linker error?

File: compiler/AST/FnSymbol.cpp
Comment (dlongnecke-cray) [reply compiler/AST/FnSymbol.cpp:348]: Sure can, thanks!

File: compiler/dyno/lib/parsing/ParserContextImpl.h
Comment (DanilaFe) [inline compiler/dyno/lib/parsing/ParserContextImpl.h:794]: This function is called `findErrorInFnBodyOrFormals`, but I only see it checking the `fp.formals`. Does `fp.formals` contain the body somehow, or is the function name slightly inaccurate?

File: compiler/dyno/lib/parsing/ParserContextImpl.h
Comment (DanilaFe) [inline compiler/dyno/lib/parsing/ParserContextImpl.h:808]: Do we want to address this before merging the PR, or do you think it's fine to get this working later?

File: compiler/dyno/lib/uast/AnonFormal.cpp
Comment (DanilaFe) [inline compiler/dyno/lib/uast/AnonFormal.cpp:35]: I think `std::unique_ptr` has an overload for `!=` and `nullptr_t`:

```C++
template <class T, class D> bool operator!= (const unique_ptr<T,D>& lhs, nullptr_t) noexcept;
```

So you might not need the `.get()` here (and maybe other places that need to compare against null)

File: compiler/dyno/lib/uast/post-parse-checks.cpp
Comment (DanilaFe) [inline compiler/dyno/lib/uast/post-parse-checks.cpp:531]: Ah, found it!

File: compiler/dyno/lib/uast/post-parse-checks.cpp
Comment (DanilaFe) [inline compiler/dyno/lib/uast/post-parse-checks.cpp:742]: So is this allowing for modules called, like, `true.chpl`? What are the reasons this is included?

**Review by DanilaFe — COMMENTED** [review]
More comments focusing on the Dyno portion. Here everything looks good, but I can't say that I fully understand how the parser context stuff works (and thus, what the implications of the changes are), so maybe another set of eyes there would be helpful.

File: compiler/dyno/include/chpl/uast/Function.h
Comment (dlongnecke-cray) [reply compiler/dyno/include/chpl/uast/Function.h:381]: I personally think it's OK to have these things return `const char*` when each enum converts to a fixed length string. E.g., within dyno I'd expect we could have chpldoc call it, (later, possibly error messages).

I'm actually in the opposite camp in that I'd find a `std::string` to be more reprehensible, because IMO it's more of a heavyweight type than is actually needed.

W.r.t. the production compiler, I wouldn't expect this to be called there because it's a different IR at that point (I know we have plans to be able to bridge that gap and inspect using dyno but I don't think we're at that point yet).

File: compiler/dyno/lib/parsing/ParserContextImpl.h
Comment (dlongnecke-cray) [reply compiler/dyno/lib/parsing/ParserContextImpl.h:808]: I think it's fine to get it working later. The whole effort should be considered `--warn-unstable`.

File: compiler/dyno/lib/uast/AnonFormal.cpp
Comment (dlongnecke-cray) [reply compiler/dyno/lib/uast/AnonFormal.cpp:35]: Ah, ok, ty!

File: compiler/dyno/lib/uast/post-parse-checks.cpp
Comment (dlongnecke-cray) [reply compiler/dyno/lib/uast/post-parse-checks.cpp:742]: In fact, after discussion I think right now we should be skipping all modules. See: https://github.com/chapel-lang/chapel/issues/19197

Feel free to cast your vote there, personally I voted to ban any module/file name that conflicts with a reserved word.

File: compiler/dyno/include/chpl/uast/Function.h
Comment (benharsh) [reply compiler/dyno/include/chpl/uast/Function.h:381]: I think I'd prefer that this method, and others like it, return a ``std::string``. I wondered at first if the null case was useful, but we can always create an exhaustive switch-statement for an enum like this.

Anyways, I think we could address this in a follow-up PR.

**Review by benharsh — APPROVED** [review]
The production compiler changes look good to me. As you indicate, there's some future work to be done that could clean things up, but this looks like a good step forward.

Could you run some testing (primers, at least) with ``--verify``, just to make sure something doesn't explode?

File: compiler/dyno/include/chpl/uast/Function.h
Comment (dlongnecke-cray) [reply compiler/dyno/include/chpl/uast/Function.h:381]: Adjusted to return `std::string`.

File: compiler/dyno/lib/parsing/ParserContextImpl.h
Comment (dlongnecke-cray) [reply compiler/dyno/lib/parsing/ParserContextImpl.h:794]: Renamed it

**Pull request merged**
